
# Hands on Data Analysis Follow Along

### Previously we learned how to import and clean data in data frames using pandas. Now we're going to learn how to operate on that data to pull out useful information.

#### We will start by learning some of the built in functions Pandas has to offer

In [2]:
#first things first:
import numpy as np
import pandas as pd

We will practice on data from College Board collected in 2010, about the average SAT scores for different schools:

In [5]:
sat = pd.read_csv('SAT_2010.csv')

In [6]:
sat #see what the Data set contains

DBN                                        School Name  \
0    01M292     Henry Street School for International Studies    
1    01M448               University Neighborhood High School    
2    01M450                   East Side Community High School    
3    01M458                      SATELLITE ACADEMY FORSYTH ST    
4    01M509                                  CMSP HIGH SCHOOL    
..      ...                                                ...   
455  75R025                        South Richmond High School    
456  75X012                      PS12X LEWIS AND CLARK SCHOOL    
457  75X754  P754 X - Jeffrey M. Rapport School for Career ...   
458  76K460                              John Jay High School    
459  79X490                                    PHOENIX SCHOOL    

     Number of Test Takers  Critical Reading Mean  Mathematics Mean  \
0                     31.0                  391.0             425.0   
1                     60.0                  394.0             419.0   
2                     69.0                  418.0             431.0   
3                     26.0                  385.0             370.0   
4                      NaN                    NaN               NaN   
..                     ...                    ...               ...   
455                   10.0                  407.0             421.0   
456                    NaN                    NaN               NaN   
457                    NaN                    NaN               NaN   
458                    9.0                  390.0             381.0   
459                    7.0                  404.0             423.0   

     Writing Mean  
0           385.0  
1           387.0  
2           402.0  
3           378.0  
4             NaN  
..            ...  
455         400.0  
456           NaN  
457           NaN  
458         398.0  
459         416.0  

[460 rows x 6 columns]

If we only want to see the first 'n' items of the table, where n is some positive integer, we can use the .head(n) function.

To display only the first 3 items of our data frame 'sat':

In [7]:
sat.head(3)

DBN                                     School Name  \
0  01M292  Henry Street School for International Studies    
1  01M448            University Neighborhood High School    
2  01M450                East Side Community High School    

   Number of Test Takers  Critical Reading Mean  Mathematics Mean  \
0                   31.0                  391.0             425.0   
1                   60.0                  394.0             419.0   
2                   69.0                  418.0             431.0   

   Writing Mean  
0         385.0  
1         387.0  
2         402.0

Now, let's use the function .info(), this prints out information of a Data Frame, seeing how it works is best understood by seeing it's output:

In [8]:
#now use the .info() function, and see what information it gives you:
sat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 460 entries, 0 to 459
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   DBN                    460 non-null    object 
 1   School Name            460 non-null    object 
 2   Number of Test Takers  386 non-null    float64
 3   Critical Reading Mean  386 non-null    float64
 4   Mathematics Mean       386 non-null    float64
 5   Writing Mean           386 non-null    float64
dtypes: float64(4), object(2)
memory usage: 21.7+ KB


Now we can easily see what each column is, and how much data it contains. What makes some columns contain 460 pieces of information while others contain only 386?

Let's try the function .isnull(), which returns the data frame back with Boolean Values, depending on if a value is null (NaN) or not:

In [9]:
sat.isnull()

DBN  School Name  Number of Test Takers  Critical Reading Mean  \
0    False        False                  False                  False   
1    False        False                  False                  False   
2    False        False                  False                  False   
3    False        False                  False                  False   
4    False        False                   True                   True   
..     ...          ...                    ...                    ...   
455  False        False                  False                  False   
456  False        False                   True                   True   
457  False        False                   True                   True   
458  False        False                  False                  False   
459  False        False                  False                  False   

     Mathematics Mean  Writing Mean  
0               False         False  
1               False         False  
2               False         False  
3               False         False  
4                True          True  
..                ...           ...  
455             False         False  
456              True          True  
457              True          True  
458             False         False  
459             False         False  

[460 rows x 6 columns]

The miss-match in data quantity from column to column comes from some of the values being Null.

Let's consider another possible issue, how can we tell if certain data has repeats? We can use the function .nunique() to see how many unique values are in each column:

In [10]:
sat.nunique()

DBN                      458
School Name              460
Number of Test Takers    162
Critical Reading Mean    157
Mathematics Mean         165
Writing Mean             147
dtype: int64

The number of DBN values (an arbitrary code given to each school) is less than the number of unique school names, what does this mean and why may it be problematic for School Board?

How can we find which DBN is repeated?

We can use the function .value_counts(), however we need to use it only on a specific column, which we do using the same syntax we learned in for data cleaning:

In [6]:
sat['DBN'].value_counts()

10X430    2
04M610    2
01M450    1
15K464    1
24Q530    1
         ..
17K470    1
24Q485    1
19K615    1
09X241    1
10X374    1
Name: DBN, Length: 458, dtype: int64

Now we can see there are two DBN Values that repeat. Notice that the .value_counts() function automatically sorts in order from greatest to least.

Practice pulling the data from the data frame where the DBN is one of the repeated values to see what schools share their values:

In [14]:
#sat['DBN'].value_counts('10X430')
sat[(sat['DBN']=='10X430') | (sat['DBN']== '04M610')]

DBN                                  School Name  \
86   04M610               YOUNG WOMEN LEADERSHIP SCHOOL    
87   04M610  Young Women's Leadership School of Astoria    
176  10X430                          WALTON HIGH SCHOOL    
177  10X430                  YABC AT WALTON HIGH SCHOOL    

     Number of Test Takers  Critical Reading Mean  Mathematics Mean  \
86                    48.0                  451.0             445.0   
87                     NaN                    NaN               NaN   
176                    NaN                    NaN               NaN   
177                    NaN                    NaN               NaN   

     Writing Mean  
86          470.0  
87            NaN  
176           NaN  
177           NaN

The repeated values are given to what looks like two branches of the same two schools, however the extra repeats all have null values, so they won't scew the data in any way if we remove them.

Clear out all the schools (rows) that contain Null Data using the methods learned in the Data Cleaning lesson, make sure to reset the index as well. (Do this in-place to permanenetly alter the Data Frame)

In [21]:
sat.dropna(axis = 0, inplace=True)
sat.reset_index(drop=True,inplace = True)
sat

DBN                                       School Name  \
0    01M292    Henry Street School for International Studies    
1    01M448              University Neighborhood High School    
2    01M450                  East Side Community High School    
3    01M458                     SATELLITE ACADEMY FORSYTH ST    
4    01M515          Lower East Side Preparatory High School    
..      ...                                               ...   
381  32K554                    ALL CITY LEADERSHIP SECONDARY    
382  32K556  Bushwick Leaders High School for Academic Excel    
383  75R025                       South Richmond High School    
384  76K460                             John Jay High School    
385  79X490                                   PHOENIX SCHOOL    

     Number of Test Takers  Critical Reading Mean  Mathematics Mean  \
0                     31.0                  391.0             425.0   
1                     60.0                  394.0             419.0   
2                     69.0                  418.0             431.0   
3                     26.0                  385.0             370.0   
4                    154.0                  314.0             532.0   
..                     ...                    ...               ...   
381                   29.0                  394.0             420.0   
382                   30.0                  357.0             345.0   
383                   10.0                  407.0             421.0   
384                    9.0                  390.0             381.0   
385                    7.0                  404.0             423.0   

     Writing Mean  
0           385.0  
1           387.0  
2           402.0  
3           378.0  
4           314.0  
..            ...  
381         395.0  
382         351.0  
383         400.0  
384         398.0  
385         416.0  

[386 rows x 6 columns]

In [22]:
sat #Does the result look correct?

DBN                                       School Name  \
0    01M292    Henry Street School for International Studies    
1    01M448              University Neighborhood High School    
2    01M450                  East Side Community High School    
3    01M458                     SATELLITE ACADEMY FORSYTH ST    
4    01M515          Lower East Side Preparatory High School    
..      ...                                               ...   
381  32K554                    ALL CITY LEADERSHIP SECONDARY    
382  32K556  Bushwick Leaders High School for Academic Excel    
383  75R025                       South Richmond High School    
384  76K460                             John Jay High School    
385  79X490                                   PHOENIX SCHOOL    

     Number of Test Takers  Critical Reading Mean  Mathematics Mean  \
0                     31.0                  391.0             425.0   
1                     60.0                  394.0             419.0   
2                     69.0                  418.0             431.0   
3                     26.0                  385.0             370.0   
4                    154.0                  314.0             532.0   
..                     ...                    ...               ...   
381                   29.0                  394.0             420.0   
382                   30.0                  357.0             345.0   
383                   10.0                  407.0             421.0   
384                    9.0                  390.0             381.0   
385                    7.0                  404.0             423.0   

     Writing Mean  
0           385.0  
1           387.0  
2           402.0  
3           378.0  
4           314.0  
..            ...  
381         395.0  
382         351.0  
383         400.0  
384         398.0  
385         416.0  

[386 rows x 6 columns]

In [23]:
sat.info() #make sure each column now only contains 386 pieces of data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 386 entries, 0 to 385
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   DBN                    386 non-null    object 
 1   School Name            386 non-null    object 
 2   Number of Test Takers  386 non-null    float64
 3   Critical Reading Mean  386 non-null    float64
 4   Mathematics Mean       386 non-null    float64
 5   Writing Mean           386 non-null    float64
dtypes: float64(4), object(2)
memory usage: 18.2+ KB


Similar to .nunique() is the function .unique(), this function is applied to a single column at a time, using the same syntax as .value_counts().

We will use a data set of employee data for this:

In [29]:
HR_Data = pd.read_csv('HR.csv')

In [35]:
HR_Data.head(10) #get a look at what data this contains

Employee_Name       EmpID  PayRate                  Position State  \
0            Brown, Mia  1103024456    28.50              Accountant I    MA   
1  LaRotonda, William    1106026572    23.00              Accountant I    MA   
2      Steans, Tyrone    1302053333    29.00              Accountant I    MA   
3       Howard, Estelle  1211050782    21.50  Administrative Assistant    MA   
4           Singh, Nan   1307059817    16.56  Administrative Assistant    MA   
5      Smith, Leigh Ann   711007713    20.50  Administrative Assistant    MA   
6      Bunbury, Jessica  1504073368    55.00        Area Sales Manager    VA   
7     Carter, Michelle   1403065721    55.00        Area Sales Manager    VT   
8     Dietrich, Jenna    1408069481    55.00        Area Sales Manager    WA   
9      Digitale, Alfred  1306059197    56.00        Area Sales Manager    NH   

     Zip Sex MaritalDesc  DateofHire        EmploymentStatus     Department  \
0   1450   F     Married  10/27/2008                  Active  Admin Offices   
1   1460  M     Divorced    1/6/2014                  Active  Admin Offices   
2   2703  M       Single   9/29/2014                  Active  Admin Offices   
3   2170   F     Married   2/16/2015    Terminated for Cause  Admin Offices   
4   2330   F      Single    5/1/2015                  Active  Admin Offices   
5   1844   F     Married   9/26/2011  Voluntarily Terminated  Admin Offices   
6  21851   F     Married   8/15/2011  Voluntarily Terminated          Sales   
7   5664   F      Single   8/18/2014                  Active          Sales   
8  98052   F      Single   2/20/2012                  Active          Sales   
9   3062  M      Married   8/18/2014                  Active          Sales   

          ManagerName  EngagementSurvey  EmpSatisfaction  \
0  Brandon R. LeBlanc              2.04                2   
1  Brandon R. LeBlanc              5.00                4   
2  Brandon R. LeBlanc              3.90                5   
3  Brandon R. LeBlanc              3.24                3   
4  Brandon R. LeBlanc              5.00                3   
5  Brandon R. LeBlanc              3.80                4   
6          John Smith              3.14                5   
7          John Smith              5.00                5   
8          John Smith              2.30                1   
9          John Smith              3.60                5   

   SpecialProjectsCount  DaysLateLast30  
0                     6             0.0  
1                     4             0.0  
2                     5             0.0  
3                     4             NaN  
4                     5             0.0  
5                     4             NaN  
6                     0             NaN  
7                     0             0.0  
8                     0             0.0  
9                     0             0.0

Use .unique() on the column 'Position' to find the different jobs that exist in the data set:

In [40]:
HR_Data['Position'].unique()

array(['Accountant I', 'Administrative Assistant', 'Area Sales Manager',
       'BI Developer', 'BI Director', 'CIO', 'Data Architect',
       'Database Administrator', 'Data Analyst', 'Data Analyst ',
       'Director of Operations', 'Director of Sales', 'IT Director',
       'IT Manager - DB', 'IT Manager - Infra', 'IT Manager - Support',
       'IT Support', 'Network Engineer', 'President & CEO',
       'Production Manager', 'Production Technician I',
       'Production Technician II', 'Sales Manager', 'Senior BI Developer',
       'Shared Services Manager', 'Software Engineer',
       'Software Engineering Manager', 'Sr. Accountant', 'Sr. DBA',
       'Enterprise Architect', 'Principal Data Architect',
       'Sr. Network Engineer'], dtype=object)

.nunique() can also be used on a single column, given is one way of finding the number of unique jobs by using .unique(), but try using .nunique() to find the value:

In [37]:
len(HR_Data['Position'].unique())

32

In [38]:
#find the same information with nunique()
HR_Data['Position'].nunique()

32

Some data sets have a lot of columns, rather than manually scrolling through the ouput of .head() or .info() we could simply use .columns to return a list of strings

In [41]:
HR_Data.columns 
#Note: we do not use (), this is not a function, but an attribute of the data frame object, just like 'shape'

Index(['Employee_Name', 'EmpID', 'PayRate', 'Position', 'State', 'Zip', 'Sex',
       'MaritalDesc', 'DateofHire', 'EmploymentStatus', 'Department',
       'ManagerName', 'EngagementSurvey', 'EmpSatisfaction',
       'SpecialProjectsCount', 'DaysLateLast30'],
      dtype='object')

We can sort the data based off of a column using .sort_values('column name')

Try sorting HR_Data by EmpID:

In [42]:
HR_Data.sort_values('EmpID')

Employee_Name       EmpID  PayRate                  Position  \
62      Lindsay, Leonara    602000312     26.0                IT Support   
166     Dickinson, Geoff    706006285     21.0   Production Technician I   
216  Kinsella, Kathleen     710007401     22.0   Production Technician I   
129        Rose, Ashley     710007555     17.0   Production Technician I   
5        Smith, Leigh Ann   711007713     20.5  Administrative Assistant   
..                    ...         ...      ...                       ...   
184      Jacobi, Hannah    1503072857     21.0   Production Technician I   
20           Buck, Edward  1504073313     55.0        Area Sales Manager   
6        Bunbury, Jessica  1504073368     55.0        Area Sales Manager   
120         Cockel, James  1599991009     15.0   Production Technician I   
67        Gonzalez, Maria  1988299991     39.0                IT Support   

    State    Zip Sex MaritalDesc DateofHire        EmploymentStatus  \
62     CT   6070   F      Single  1/21/2011                  Active   
166    MA   2180  M       Single  5/12/2014                  Active   
216    MA   2170   F     Married  9/26/2011  Voluntarily Terminated   
129    MA   1886   F   Separated   1/6/2014                  Active   
5      MA   1844   F     Married  9/26/2011  Voluntarily Terminated   
..    ...    ...  ..         ...        ...                     ...   
184    MA   1778   F    Divorced  9/30/2013                  Active   
20     MA   2330  M      Married  9/29/2014                  Active   
6      VA  21851   F     Married  8/15/2011  Voluntarily Terminated   
120    MA   2452  M       Single   7/8/2013                  Active   
67     MA   2472   F   Separated   1/5/2015                  Active   

            Department         ManagerName  EngagementSurvey  EmpSatisfaction  \
62               IT/IS        Eric Dougall              4.64                4   
166  Production              Ketsia Liebig              1.13                4   
216  Production             Webster Butler              3.60                5   
129  Production              David Stanley              5.00                4   
5        Admin Offices  Brandon R. LeBlanc              3.80                4   
..                 ...                 ...               ...              ...   
184  Production             Kissy Sullivan              2.05                5   
20               Sales       Lynn Daneault              3.31                5   
6                Sales          John Smith              3.14                5   
120  Production              David Stanley              1.51                5   
67               IT/IS        Peter Monroe              4.61                4   

     SpecialProjectsCount  DaysLateLast30  
62                      5             0.0  
166                     0             0.0  
216                     0             NaN  
129                     0             0.0  
5                       4             NaN  
..                    ...             ...  
184                     0             0.0  
20                      0             0.0  
6                       0             NaN  
120                     0             0.0  
67                      5             0.0  

[310 rows x 16 columns]

We can use a variety of other functions on single columns to find information about them, such as .max(), .min(), .mean(), .median(), .sum(), and more using the same syntax we did for .unique() or .value_counts(), lets practice:

In [43]:
#What's the average pay rate?
HR_Data['PayRate'].mean()

31.28480645161292

In [46]:
#in the SAT data, what was the lowest average writing score for a school?
#your code here
#sat.sort_values('Writing Mean')
sat['Writing Mean'].min()

285.0

But what school does this data belong to? We can use the Syntax we did for cleaning to pull that specific row:

In [ ]:
sat[sat['Writing Mean'] == sat['Writing Mean'].min()]

Find the name of the employee in our HR_Data with the Highest Pay

In [48]:
HR_Data['PayRate'].max()

80.0

### Practice

The toal SAT score is calculated by simply adding up the scores for reading, writing, and math. Lets try to get a weighted average of the total SAT score for all the students. We say 'weighted average' because each school had a different number of students who contributed to it's average score. 

To get the weighted average, first, find the avearge total SAT score for each school by summing the average of math, reading, and writing. Then, multiply this number by the number of students that contributed to the schools score. Take the sum of this product from all schools, and divide it by the sum of all students from all schools.

Challenege: do this all in one line

Ans: 1335.922

In [57]:
sat['Weighted Average'] = sat['Critical Reading Mean']+sat['Mathematics Mean']+ sat['Writing Mean']
#tips['tip_percent']=tips['tip']/tips['total_bill']
#sum(sat['Weighted Average']/386)
sum(sat['Weighted Average']*sat['Number of Test Takers']/sat['Number of Test Takers'].sum())

1335.921823452963

### Correlations in Data with .corr()

The last built in function we'll look at is .corr()

This function finds the correlation between the data in the columns of a data frame, and outputs them into a matrix

In [58]:
sat.corr()

Number of Test Takers  Critical Reading Mean  \
Number of Test Takers               1.000000               0.439738   
Critical Reading Mean               0.439738               1.000000   
Mathematics Mean                    0.547275               0.894079   
Writing Mean                        0.450994               0.978404   
Weighted Average                    0.495145               0.980187   

                       Mathematics Mean  Writing Mean  Weighted Average  
Number of Test Takers          0.547275      0.450994          0.495145  
Critical Reading Mean          0.894079      0.978404          0.980187  
Mathematics Mean               1.000000      0.902777          0.960161  
Writing Mean                   0.902777      1.000000          0.983536  
Weighted Average               0.960161      0.983536          1.000000

Notice that the rows and columns of this output are the same.

Each value in the matrix will have a value between -1 and 1. A value of 0 means that the two sets of data have no correlation aka they're unrelated/independendt of each other. A value of 1 means that two sets of Data are Completely coorelated, aka equal to each other. A negative correlation means two data sets are related in that an increase in one means a decrease in the other.

Take a look at the correlation within the HR_Data data frame:

In [59]:
HR_Data.corr()

EmpID   PayRate       Zip  EngagementSurvey  \
EmpID                 1.000000 -0.020310  0.130735         -0.005720   
PayRate              -0.020310  1.000000  0.391945         -0.024223   
Zip                   0.130735  0.391945  1.000000         -0.000982   
EngagementSurvey     -0.005720 -0.024223 -0.000982          1.000000   
EmpSatisfaction      -0.017726  0.061525 -0.064478         -0.017123   
SpecialProjectsCount -0.171329  0.448484 -0.096083         -0.040253   
DaysLateLast30             NaN       NaN       NaN               NaN   

                      EmpSatisfaction  SpecialProjectsCount  DaysLateLast30  
EmpID                       -0.017726             -0.171329             NaN  
PayRate                      0.061525              0.448484             NaN  
Zip                         -0.064478             -0.096083             NaN  
EngagementSurvey            -0.017123             -0.040253             NaN  
EmpSatisfaction              1.000000              0.036456             NaN  
SpecialProjectsCount         0.036456              1.000000             NaN  
DaysLateLast30                    NaN                   NaN             NaN

As a disclaimer, this data is made up for the purposes of practicing data science.

One way to confirm this is to look at the coorelation of the zip code of the employees to other data. Any given Zip Code should be nearly random for every employee, as it's arbitrarily based on location. However we see a somewhat significant correlation between the zip code and the pay rate. Why is this? Perhaps the PayRate was generated based off of the zip code when the data set was made, or vis-versa. Perhaps there is a legitamate reason for it, maybe people in a certain location have a higher pay-rates due to a building with executives residing in a region with a higher zip code, while buildings with lower level employees reside in regions of smaller zip codes.

The math that generates the output comes from statistics. We don't have time to prove it, but the general formula is:

$ \rho_{X,Y} = \frac{E[XY]-E[X]E[Y]}{\sqrt{E[X^2]-[E[X]]^2} \sqrt{E[Y^2]-[E[Y]]^2}} $
 
 Where $\rho_{X,Y}$ is the correlation value of data sets (Columns) X and Y, $E[X]$ is the 'expectation value' (fancy term for mean) of the values of X, $E[Y]$ is the 'expectation value' of the values of Y, and likewise $E[XY]$ would be the average of every corresponding data point x and y multiplied together, and similar for $E[X^2], E[Y^2]$
 
If you don't understand this yet it's okay, you woun't need to program this in yourself, just know that the numbers generated by corr() is mathmatics and not magic

### Apply() and user defined/ Lambda functions

Lets look again at our HR_Data:

In [60]:
HR_Data.head(3)

Employee_Name       EmpID  PayRate      Position State   Zip Sex  \
0            Brown, Mia  1103024456     28.5  Accountant I    MA  1450   F   
1  LaRotonda, William    1106026572     23.0  Accountant I    MA  1460  M    
2      Steans, Tyrone    1302053333     29.0  Accountant I    MA  2703  M    

  MaritalDesc  DateofHire EmploymentStatus     Department         ManagerName  \
0     Married  10/27/2008           Active  Admin Offices  Brandon R. LeBlanc   
1    Divorced    1/6/2014           Active  Admin Offices  Brandon R. LeBlanc   
2      Single   9/29/2014           Active  Admin Offices  Brandon R. LeBlanc   

   EngagementSurvey  EmpSatisfaction  SpecialProjectsCount  DaysLateLast30  
0              2.04                2                     6             0.0  
1              5.00                4                     4             0.0  
2              3.90                5                     5             0.0

Say we want to make a column of data that contained only the year an employee was hired, how can we extract this data from the 'DateofHire' column? 

This is where the .apply() function comes in, .apply() takes in a function as it's argument, and applies it to every value within the series (either a row or a column).

Maybe we could find a built in function to do this, but it's far easier to write a function ourselves to do this for our specific formatting.

Lets first define a function that uses split('/') on the 'DateOfHire' Column to get a list of \[month, day, year\], to ultimately return just the year:

In [61]:
def year_extract(date_string):
    return date_string.split('/')[2]

In [62]:
#test it out
test_data = '2/3/2021'
print(year_extract(test_data))

2021


Now lets apply the function to create a new column that contains just the year the employee was hired:

In [63]:
HR_Data['YearofHire'] = HR_Data['DateofHire'].apply(year_extract)
#we don't need to pass in any arguments to our function 'year_extract', the data in the row is assumed to be the argument

In [64]:
HR_Data.head(4)

Employee_Name       EmpID  PayRate                  Position State  \
0            Brown, Mia  1103024456     28.5              Accountant I    MA   
1  LaRotonda, William    1106026572     23.0              Accountant I    MA   
2      Steans, Tyrone    1302053333     29.0              Accountant I    MA   
3       Howard, Estelle  1211050782     21.5  Administrative Assistant    MA   

    Zip Sex MaritalDesc  DateofHire      EmploymentStatus     Department  \
0  1450   F     Married  10/27/2008                Active  Admin Offices   
1  1460  M     Divorced    1/6/2014                Active  Admin Offices   
2  2703  M       Single   9/29/2014                Active  Admin Offices   
3  2170   F     Married   2/16/2015  Terminated for Cause  Admin Offices   

          ManagerName  EngagementSurvey  EmpSatisfaction  \
0  Brandon R. LeBlanc              2.04                2   
1  Brandon R. LeBlanc              5.00                4   
2  Brandon R. LeBlanc              3.90                5   
3  Brandon R. LeBlanc              3.24                3   

   SpecialProjectsCount  DaysLateLast30 YearofHire  
0                     6             0.0       2008  
1                     4             0.0       2014  
2                     5             0.0       2014  
3                     4             NaN       2015

Now try writing two functions that takes the employee_Name and splits it into the first and last names respectivelys:

In [82]:
def extract_last(name):
    return name.split(',')[1]

def extract_first(name):
    return name.split(',')[0]

Now apply this function to add two separate columns of 'FirstName' and 'LastName':

In [83]:
test_data = 'Tyler, MacBean'
print(extract_first(test_data))
print(extract_last(test_data))

Tyler
 MacBean


In [84]:
HR_Data['First_Name'] = HR_Data['Employee_Name'].apply(extract_first)
HR_Data['Last_Name'] = HR_Data['Employee_Name'].apply(extract_last)
HR_Data

Employee_Name       EmpID  PayRate                  Position  \
0                Brown, Mia  1103024456    28.50              Accountant I   
1      LaRotonda, William    1106026572    23.00              Accountant I   
2          Steans, Tyrone    1302053333    29.00              Accountant I   
3           Howard, Estelle  1211050782    21.50  Administrative Assistant   
4               Singh, Nan   1307059817    16.56  Administrative Assistant   
..                      ...         ...      ...                       ...   
305          Daniele, Ann    1411071312    54.10      Sr. Network Engineer   
306          Lajiri,  Jyoti  1108028108    56.20      Sr. Network Engineer   
307  Semizoglou, Jeremiah     904013591    53.80      Sr. Network Engineer   
308              South, Joe  1308060959    53.00      Sr. Network Engineer   
309         Warfield, Sarah  1301052347    55.20      Sr. Network Engineer   

    State   Zip Sex MaritalDesc  DateofHire      EmploymentStatus  \
0      MA  1450   F     Married  10/27/2008                Active   
1      MA  1460  M     Divorced    1/6/2014                Active   
2      MA  2703  M       Single   9/29/2014                Active   
3      MA  2170   F     Married   2/16/2015  Terminated for Cause   
4      MA  2330   F      Single    5/1/2015                Active   
..    ...   ...  ..         ...         ...                   ...   
305    CT  6033   F     Married  11/10/2014      Leave of Absence   
306    MA  2169  M      Married  11/10/2014      Leave of Absence   
307    MA  1545  M      Married   6/30/2016          Future Start   
308    CT  6278  M       Single  11/10/2014                Active   
309    MA  2453   F     Widowed   3/30/2015                Active   

        Department         ManagerName  EngagementSurvey  EmpSatisfaction  \
0    Admin Offices  Brandon R. LeBlanc              2.04                2   
1    Admin Offices  Brandon R. LeBlanc              5.00                4   
2    Admin Offices  Brandon R. LeBlanc              3.90                5   
3    Admin Offices  Brandon R. LeBlanc              3.24                3   
4    Admin Offices  Brandon R. LeBlanc              5.00                3   
..             ...                 ...               ...              ...   
305          IT/IS        Peter Monroe              1.55                5   
306          IT/IS        Peter Monroe              4.46                5   
307          IT/IS        Peter Monroe              3.27                4   
308          IT/IS        Peter Monroe              5.00                3   
309          IT/IS        Peter Monroe              1.06                4   

     SpecialProjectsCount  DaysLateLast30 YearofHire  First_Name  \
0                       6             0.0       2008       Brown   
1                       4             0.0       2014   LaRotonda   
2                       5             0.0       2014      Steans   
3                       4             NaN       2015      Howard   
4                       5             0.0       2015       Singh   
..                    ...             ...        ...         ...   
305                     8             0.0       2014     Daniele   
306                     6             0.0       2014      Lajiri   
307                     5             0.0       2016  Semizoglou   
308                     4             0.0       2014       South   
309                     8             0.0       2015    Warfield   

    YearofHireLambda    Last_Name  
0               2008          Mia  
1               2014    William    
2               2014     Tyrone    
3               2015      Estelle  
4               2015         Nan   
..               ...          ...  
305             2014        Ann    
306             2014        Jyoti  
307             2016   Jeremiah    
308             2014          Joe  
309             2015        Sarah  

[310 rows x 20 columns]

In [79]:
HR_Data.head(3) #check your output

Employee_Name       EmpID  PayRate      Position State   Zip Sex  \
0            Brown, Mia  1103024456     28.5  Accountant I    MA  1450   F   
1  LaRotonda, William    1106026572     23.0  Accountant I    MA  1460  M    
2      Steans, Tyrone    1302053333     29.0  Accountant I    MA  2703  M    

  MaritalDesc  DateofHire EmploymentStatus     Department         ManagerName  \
0     Married  10/27/2008           Active  Admin Offices  Brandon R. LeBlanc   
1    Divorced    1/6/2014           Active  Admin Offices  Brandon R. LeBlanc   
2      Single   9/29/2014           Active  Admin Offices  Brandon R. LeBlanc   

   EngagementSurvey  EmpSatisfaction  SpecialProjectsCount  DaysLateLast30  \
0              2.04                2                     6             0.0   
1              5.00                4                     4             0.0   
2              3.90                5                     5             0.0   

  YearofHire First_Name  
0       2008      Brown  
1       2014  LaRotonda  
2       2014     Steans

This method of defining a function and then using it works just fine, but it can be a bit cluncky. Because we only use them once, we can abreviate it in our code by passing in a 'Lambda Function', which is the same thing we just did, but we don't give our function a name or give it permanent memory space, but instead breifly pass in the formula to be used the one time.

The syntax for them is different than what we've seen, so let's start by re-creating the year_extract function and break-down the differences:

In [85]:
HR_Data['YearofHireLambda'] = HR_Data['DateofHire'].apply(lambda year: year.split('/')[2])
HR_Data['LambdaFirst'] = HR_Data['Employee_Name'].apply(lambda first: first.split(',')[0])

In [86]:
HR_Data.head(3)

Employee_Name       EmpID  PayRate      Position State   Zip Sex  \
0            Brown, Mia  1103024456     28.5  Accountant I    MA  1450   F   
1  LaRotonda, William    1106026572     23.0  Accountant I    MA  1460  M    
2      Steans, Tyrone    1302053333     29.0  Accountant I    MA  2703  M    

  MaritalDesc  DateofHire EmploymentStatus  ...         ManagerName  \
0     Married  10/27/2008           Active  ...  Brandon R. LeBlanc   
1    Divorced    1/6/2014           Active  ...  Brandon R. LeBlanc   
2      Single   9/29/2014           Active  ...  Brandon R. LeBlanc   

  EngagementSurvey  EmpSatisfaction  SpecialProjectsCount  DaysLateLast30  \
0             2.04                2                     6             0.0   
1             5.00                4                     4             0.0   
2             3.90                5                     5             0.0   

   YearofHire First_Name YearofHireLambda   Last_Name LambdaFirst  
0        2008      Brown             2008         Mia       Brown  
1        2014  LaRotonda             2014   William     LaRotonda  
2        2014     Steans             2014    Tyrone        Steans  

[3 rows x 21 columns]

We did the exact same thing, but all in one line!
How did the syntax translate? we wrote:

`lambda year : year.split('/')[2]`

Think of 'lambda' as our 'def', and 'year' as the name of our input variable, to give `lambda year :` to be like `def nameless_function(year):`

The left side of the colon is our function definition, see how it's the same thing we did in the original function definition

We can abbreviate further, most times when people write lambda functions they simply use 'x' and 'y' as their variables,

For example `lambda x: x.split(',')[0]` may be used to get the last names if this lambda function were applied to the 'Employee_Name' column

When working on your own code, if you don't feel comfortable with lambda functions right away it's okay. You can still use the first method that gives it an identifier. Later if you feel up to the challenge, try to go back through and adapt them to lambda functions.

### Practice

In [ ]:
HR_Data = pd.read_csv('HR.csv')
HR_Data.drop(axis=1,inplace=True,labels='Department') #modifying our table as if the department column were not included

In the HR data set, there are 5 positions relate to IT. First, try using .unique() to see the different positions in the department, then use .apply() to generate a new data frame with only people relating to the IT department by devising your own function.

Export this as a new csv file 'HR_IT_Data.csv' without the index (I'm not going to give you this answer, try using the help() function or google. Self teaching is the best way to learn past the basics, as teaching every function is impossible in our amount of time)

Challenge: Do this all in one line

In [ ]:
#your code here

In [ ]:
#check your answer
HR_IT = pd.read_csv('HR_IT_Data.csv')
HR_IT

## Group by

The last thing we will learn is 'Group-By'
This will enable us to group data by certain values, for example, we could group people in the HR_IT data by their manager name, in order to see how different managers effect employee satisfaction or engangement.

In [ ]:
HR_IT.groupby('ManagerName')

This returns an object that, by it-self, we can't pull data from, or even display. However, we can apply functions to such as .mean(), .sum(), and other functions.

In [ ]:
IT_Manager = HR_IT.groupby('ManagerName') #give our object a name

In [ ]:
IT_Manager.mean() #find the mean values across each manager group

This is powerful, We can show analytically who is performing well and who is performing poorly, all with just a few lines of code. If you were to present this to a board or an exective, you could clean it up, screen shot it, and put it in a powerpoint with an explanation of how you obtained the data. It is clear, precise, and objective.

### Practice
Try running this same process on the entrire set of HR_Data and sort by the sum of employee satisfaction and engagement:

In [ ]:
#your code here

Is there a correlation between average payrates of employees under a manager and the sum of engagement and satisfaction averages?